modules/ipc/ipcProviderBackend.js

```diff
@@ -1,670 +1,400 @@
-        if(!socket) return;
-
-        // log.info('Called ipcProvider-destroy');
-
-        if(socket) {
-            socket.destroy();
-        }
-    });
-
-
-    var sendRequest = function(event, payload, sync) {
-        log.trace('sendRequest', event.sender.getId(), payload, sync);
-
-        var socket = global.sockets['id_'+ event.sender.getId()];
-
-        if(!socket) {
-            log.trace('Create socket');
-
-            // TODO: should we really try to reconnect, after the connection was destroyed?
-            socket = global.sockets['id_'+ event.sender.getId()] = new GethConnection(event);
-        // make sure we are connected
-        } else if(!socket.ipcSocket.writable) {
-            log.trace('Ensure socket is connected');
-
-            socket.connect(event);
-        }
-
-        // if not writeable send error back
-        if(!socket.ipcSocket.writable) {
-            log.trace('Socket not writeable');
-
-            if(event.sync)
-                event.returnValue = JSON.stringify(returnError(jsonPayload, errorTimeout));
-            else
-                event.sender.send('ipcProvider-data', JSON.stringify(returnError(jsonPayload, errorTimeout)));
-            return;
-        }
-
-        // log.info('SEND REQ', event.sender.getId());
-
-        var jsonPayload = JSON.parse(payload),
-            filteredPayload = socket.filterRequestResponse(jsonPayload);
-
-
-        if(sync === true)
-            event.sync = sync;
-
-
-        // return error, if permission not passed
-        if(_.isEmpty(filteredPayload)) {
-            log.trace('Not permitted to do request');
+            e.id = item.id;
 
-            if(event.sync)
-                event.returnValue = JSON.stringify(returnError(jsonPayload, errorMethod));
-            else
-                event.sender.send('ipcProvider-data', JSON.stringify(returnError(jsonPayload, errorMethod)));
+            return e;
+        });
 
-            return;
-        }
+        return _.isArray(payload) ? e : e[0];
+    }
 
 
+    /**
+    Make the retrun response object.
 
-        socket.checkRequests(filteredPayload, event, function(e, result, returnVal){
-            log.trace('Got result', e, result);
+    @param {Object|Array} payload Original payload
+    @param {Object|Array} value Response result.
 
-            // RETURN VALUE
-            if(!e && !_.isEmpty(returnVal)) {
-                log.trace('Return value received');
+    @method makeReturnValue
+    */
+    _makeReturnValue (payload, value) {
+        value = [].concat(value);
 
-                if(event.sync)
-                    event.returnValue = JSON.stringify(returnValue(jsonPayload, returnVal));
-                else
-                    event.sender.send('ipcProvider-data', JSON.stringify(returnValue(jsonPayload, returnVal)));
+        let allResults = ([].concat(payload)).map((item, idx) => {
+            let ret = {
+                jsonrpc: "2.0"
+            };
 
+            if (value) {
```

Comment:
I assume this is what introduces the bug that sometimes results are returned without `.result` property!

Did you take batch RPC requests into account? which come in form of an array.
Or results, which are objects, or arrays, or simple strings?